### ISSUE TITLE
Knowledge Distillation 

### ISSUE THREAD

**Issue by Evanc123** [BVLC/caffe#2382]
The easiest way to extract a lot of knowledge from the training data is to learn many different models in parallel. At test time we average the predictions of all the models or of a selected subset of good models that make different errors. A big ensemble is highly redundant. It has very little knowledge per parameter. 

If we have the ensemble, we can divide the averaged logits from the ensemble by a “temperature” to get a much softer distribution. Softened outputs reveal the dark knowledge in the ensemble. For example, the ensemble may output .3 for a cow, .2 for a cat, and .0005 for a bmw, revealing that the ensemble is confident that a cow and a cat are "close". Using a high temperature in the softmax, we minimize the cross entropy with the soft targets derived from the ensemble at high temperature. 

I want to implement this method in caffe. It involves n already trained networks, lets say 5 on MNIST and 1 untrained net. The flow of the program would be feedforward on both the ensemble and the untrained net, then penalize the untrained net with a weighted cross entropy of the softmax-raised temperature of the ensemble and the hard target softmax of the ground truth, then backprop on the untrained net only. 

I know I have to: 1. Find a way to get the output of the n already trained nets in the untrained nets prototxt, 2. Modify the softmax layer so that it raises the temperature as defined by http://www.ttic.edu/dl/dark14.pdf and 3. perform a weighted loss function. 

I know how to do 3, but I need some guidance about 2, and 1 especially. Any thoughts?

**Comment by longjon** [comment]
Please ask usage and modeling questions on the [caffe-users](https://groups.google.com/forum/#!forum/caffe-users) list. Thanks!
